libptable/ptable.h

```diff
@@ -0,0 +1,88 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Partition table
+ *
+ * Copyright 2020, 2023 Phoenix Systems
+ * Author: Hubert Buczynski, Lukasz Kosinski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#ifndef _LIB_PTABLE_H_
+#define _LIB_PTABLE_H_
+
+#include <stdint.h>
+#include <sys/types.h>
+
+
+/* Structure of partition table
+ *  _________________________________________________________________________
+ * |       28 B      |                 32 B * n                |     4 B     |
+ * |-----------------|-----------------------------------------|-------------|
+ * | ptable_t header | ptable_part_t 0 | ... | ptable_part_t n | magic bytes |
+ *  -------------------------------------------------------------------------
+ * 
+ *  NOTE: data in partition table should be stored in little endian
+ */
+
+
+/* Supported partition types */
+enum { ptable_raw = 0x51, ptable_jffs2 = 0x72, ptable_meterfs = 0x75 };
+
+
+typedef struct {
+	uint8_t name[8]; /* Partition name */
+	uint32_t offset; /* Partition offset (in bytes) */
+	uint32_t size;   /* Partition size (in bytes) */
+	uint8_t type;    /* Partition type */
+
+	/* Reserved bytes */
+	union {
+		struct {
+			uint8_t pad[11];
+			uint32_t crc; /* Partition checksum */
+		};
+		uint8_t reserved[15];
+	};
+} __attribute__((packed)) ptable_part_t;
+
+
+typedef struct {
+	uint32_t count; /* Number of partitions */
+
+	/* Reserved bytes */
+	union {
+		struct {
+			uint8_t pad[20];
+			uint32_t crc; /* Header checksum */
+		};
+		uint8_t reserved[24];
+	};
+
+	ptable_part_t parts[0]; /* Partitions */
+} __attribute__((packed)) ptable_t;
+
+
+typedef struct {
+	uint32_t memsz; /* Memory size (in bytes) */
+	uint32_t blksz; /* Memory block size (in bytes, minimal erase size for MTD, block size for block device) */
+	uint32_t secsz; /* Memory sector size (in bytes, page size for NAND, sector size for NOR, block size for block device) */
+
+	/* Memory access callbacks */
+	ssize_t (*read)(off_t offs, void *buff, size_t len);
+	ssize_t (*write)(off_t offs, const void *buff, size_t len);
+} ptable_memctx_t;
+
+
+/* Reads partition table from memory, ptable buffer should have memctx->secsz size */
+extern int ptable_read(ptable_t *ptable, const ptable_memctx_t *memctx);
+
+
+/* Writes partition table to memory, ptable buffer should have memctx->secsz size */
+extern int ptable_write(ptable_t *ptable, const ptable_memctx_t *memctx);
```

Comment:
I don't really like that we demand `memctx->secsz` buffers here. It might be hard to get 4 KB of temp RAM on e.g. imxrt1064 target

Reply:
After discussion we decided to remove partition table reading and writing functionality from the library. Now the user may detect and allocate appropriate `ptable` buffer.